go/geth_adapter/adapter.go

```diff
@@ -237,6 +248,157 @@ type runContextAdapter struct {
+	case tosca.StaticCall:
+		output, returnGas, err = a.evm.StaticCall(a.contract, toAddr, parameter.Input, gas)
+	case tosca.DelegateCall:
+		toAddr = gc.Address(parameter.CodeAddress)
+		output, returnGas, err = a.evm.DelegateCall(a.contract, toAddr, parameter.Input, gas)
+	case tosca.CallCode:
+		toAddr = gc.Address(parameter.CodeAddress)
+		output, returnGas, err = a.evm.CallCode(a.contract, toAddr, parameter.Input, gas, parameter.Value.ToUint256())
+	case tosca.Create:
+		var newAddr gc.Address
+		output, newAddr, returnGas, err = a.evm.Create(a.contract, parameter.Input, gas, parameter.Value.ToUint256())
+		createdAddress = tosca.Address(newAddr)
+	case tosca.Create2:
+		var newAddr gc.Address
+		vmSalt := &uint256.Int{}
+		vmSalt.SetBytes(parameter.Salt[:])
+		output, newAddr, returnGas, err = a.evm.Create2(a.contract, parameter.Input, gas, parameter.Value.ToUint256(), vmSalt)
+		createdAddress = tosca.Address(newAddr)
+	default:
+		panic(fmt.Sprintf("unsupported call kind: %v", kind))
+	}
+
+	// revert errors are not an error in Tosca
+	if err != nil && err != geth.ErrExecutionReverted {
+		return gethToVMErrors(err, parameter.Gas)
+	}
+
+	return tosca.CallResult{
+		Output:         output,
+		GasLeft:        tosca.Gas(returnGas),
+		GasRefund:      0, // refunds of nested calls are managed by the geth EVM and this adapter
+		CreatedAddress: createdAddress,
+		Success:        err == nil,
+	}, nil
+}
+
+// The geth EVM context does not provide the needed means
+// to forward an existing read-only mode through arbitrary
+// nested calls, as it would be needed. Thus, this information
+// is encoded into the hightest bit of the gas value, which is
+// interpreted as such by the Run() function above.
+// The geth implementation itself tracks the read-only state in
+// an implementation specific interpreter internal flag, which
+// is not accessible from this context. Also, this method depends
+// on a new interpreter per transaction call (for proper) scoping
+// which is not a desired trait for Tosca interpreter implementations.
+// With this trick, this requirement is circumvented.
+func encodeReadOnlyInGas(gas uint64, recipient tosca.Address, readOnly bool) uint64 {
+	if !isPrecompiledContract(recipient) {
+		if readOnly {
+			gas += (1 << 63)
+		}
+	}
+	return gas
+}
+
+func decodeReadOnlyFromGas(depth int, readOnly bool, gas uint64) (bool, uint64) {
+	if depth > 0 {
+		readOnly = readOnly || gas >= (1<<63)
+		if gas >= (1 << 63) {
+			gas -= (1 << 63)
+		}
+	}
+	return readOnly, gas
+}
+
+func gethToVMErrors(err error, gas tosca.Gas) (tosca.CallResult, error) {
+	switch err {
+	case geth.ErrInsufficientBalance:
+		// In this case, the caller get its gas back.
+		// TODO: this seems to be a geth implementation quirk that got
+		// transferred into the LFVM implementation; this should be fixed.
+		return tosca.CallResult{
+			GasLeft: gas,
```

Comment:
if I understand this function correctly, it only returns the gas when `geth.ErrInsufficientBalance`, and all other cases it just returns `success: false}, nil`, no ? would it no bee simpler to check for that error to set gas only? I'm thinking ways to reduce the amount of returns (with the same values)

Reply:
This code checks all known error to handle unknown errors differently. Only in the case of an unknown error we return it. I tried a few things during refactoring and I think this is the most readable way.